### ISSUE TITLE
[Python] Automatic Model.add_{parameter,submodel} using introspection

### ISSUE THREAD

**Issue by vbkaisetsu** [primitiv/primitiv#65]
Introduction
------------------

This issue proposes automatic `Model.add_{parameter,submodel}` feature for Python.

Currently, Model class has methods to register parameters and sub models.
```c++
Parameter p;
Model::add_parameter("p", p);
```
By the way, Python has introspection feature, so a name of this parameter can be retrieved from a variable name.
```python
self.p = Parameter()
self.add_all_param_models()
```

Design
-----------

`add_all_param_models()` method of Model class in Python searches parameters and models that are not registered, then it registers.
```python
self.param1 = Parameter()
self.add_parameter("p", self.param1)
self.param2 = Parameter()
self.param3 = Parameter()
self.add_all_param_models()
# equivalent to:
#     self.add_parameter("param2", self.param2)
#     self.add_parameter("param3", self.param3)
```
`self.param1` is not assigned to `"param1"` because it is directly named `"p"` by the user.

If an instance of `Parameter` is assigned to two or more variables, the function throws an error because it can not identify a name:
```python
self.param1 = Parameter()
self.param2 = self.param1
self.add_all_param_models() # Error
```

When a parameter is not assigned to a variable of a `Model` instance, it is not registered because the function can not retrieve a name.
```python
self.ps = [Parameter() for i in range(5)]
self.add_all_param_models() # No effect
self.p1 = self.ps[0]
:::
self.p5 = self.ps[4]
self.add_all_param_models()
# or self.add_parameter("p1", self.ps[0])
# :::
```

**Comment by odashi** [comment]
Overall, this proposal looks good to me.
One point: the parameter name `add_all_param_models` looks kind of confusing. I guess that it would like to be, e.g.:

* `add_all_parameters_and_submodels()` (longer but correct name)
* `[auto_]configure()`, `search()`, `setup()` or similar verbs

I like the second, but because Model classes could tend to have multiple user-defined functions, we basically have to avoid frequently-used words from the base interface of the Model.

**Comment by odashi** [comment]
I guess that separating *submodel* and *parameter* would be the cause that confuses this interface. It'd be better to integrate them into one function.
E.g. Making `add_submodel()` to be able to accept also parameters, and `add_all_submodels()` searches all not-yet registered submodels/parameters and adds them to the registory. In this case, the name `submodel` should be changed into more accurate one.

**Comment by odashi** [comment]
This issue is finally solved by primitiv/primitiv-python#10.
